### DISCUSSION TITLE
PyVista as a Pre-Processor of FEA

### DISCUSSION THREAD

**Discussion by tkoyama010** [pyvista/pyvista#1460] [Ideas]
As I am an FEA software development engineer, meshing in pre-processing is as important as post-processing. As you know, PyVista is a post-processor that does most of what we want to do. But what about pre-processing? I know a that  [pygmsh](https://github.com/nschloe/pygmsh) is a great python mesh generator. But, in here, let's think if PyVista can handle meshing. I am currently verifying the following:

1. For 1D meshes. PyVista's excellent API allows us to create meshes.
1. For 2D meshes. PyVista's Delaynay [2D triangle mesh example](https://docs.pyvista.org/examples/00-load/create-tri-surface.html) is sufficient.
1. About 3D meshes. The PyVista project already has a tetgen project, which does most of what I want to do. However, they have limited License to use it for work. (I am not criticizing the free license of [tetgen](https://github.com/pyvista/tetgen). The AGPL is a great license and their license should be respected.)
1. If PyVista can create a three-dimensional mesh, the MIT license will enable all preprocessing. Maybe the [`TetMesh` class in vedo project](https://github.com/marcomusy/vedo/blob/e594d67184068bc58ded149c450403e4a58da2a6/vedo/tetmesh.py#L69) may what I want to do.

As the name suggests, PyVista is a library of post-processing for visualization. However, I feel there is a possibility that preprocessing can be handled in the same cool way. I am not familiar with the VTK's meshing function. I am very happy if I could hear the information about VTK meshing. Feel free to talk about this topic.

**Comment by akaszynski** [comment]
I also use pyvista heavily in post-processing in FEA.  Generally, when I create finite element meshes, I use other meshers (either open or closed source) and represent the data in VTK formats through `pyvista`.  There's a closed source mesh metamorphosis tool that is sorta a "pre-processor" in such that it updates the coordinates of meshes and has pyvista to facilitate the data.  However, aside from basic 2D models like [PyVista Mesh Integration](https://mapdldocs.pyansys.com/examples/00-mapdl-examples/pyvista_mesh.html#sphx-glr-examples-00-mapdl-examples-pyvista-mesh-py), pyvista isn't great at generating conformal 3D meshes from CAD.

One thing to consider would be plugging into an existing open-source mesher that has great plug-ins for CAD and use `pyvista` as the visualization front-end and data backend.

**Reply by tkoyama010** [reply]
Thanks for your comment. Pyansys inspiring me a lot using Python in FEA.

**Reply by akaszynski** [reply]
We're getting there.  MAPDL is a pre and post processor, so it has a lot of potential.  APIs aren't great at the moment, but the plan is to improve support as needed to support end-to-end workflows.

**Reply by tkoyama010** [reply]
Great. FYI As far as I know, [GetFEM](http://getfem.org/tutorial/wheel.html) is the best design for the Python API for FEA.
